### PR TITLE
feat: add common_js_full_require_dependency

### PR DESCRIPTION
**Pull request by LingyuCoder**
## Summary

Add CommonJsFullRequireDependency

## Test Plan

- Add `packages/rspack/tests/cases/cjs-tree-shaking/full-require`

## Require Documentation?

- [x] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

File: packages/rspack/tests/__snapshots__/Stats.test.ts.snap [hunk 1]
```diff
@@ -122,6 +122,13 @@ exports[`Stats should have stats 1`] = `
               "type": "entry",
               "userRequest": "./fixtures/a",
             },
+            {
+              "moduleId": "585",
+              "moduleIdentifier": "<PROJECT_ROOT>/tests/fixtures/a.js",
+              "moduleName": "./fixtures/a.js",
+              "type": "cjs self exports reference",
```
Comment (LingyuCoder) [inline packages/rspack/tests/__snapshots__/Stats.test.ts.snap:129]: Webpack also have cjs self exports reference info in stats

File: crates/rspack_core/src/self_module_factory.rs [hunk 2]
```diff
@@ -12,8 +12,6 @@ impl ModuleFactory for SelfModuleFactory {
     let issuer = data
       .issuer_identifier
       .expect("self module must have issuer");
-    Ok(ModuleFactoryResult::new_with_module(Box::new(
-      SelfModule::new(issuer),
-    )))
+    Ok(ModuleFactoryResult::new_with_module(Box::new(SelfModule::new(issuer))).from_cache(true))
```
Comment (LingyuCoder) [inline crates/rspack_core/src/self_module_factory.rs:15]: This is a proxy module which no need to be resolved, so always from_cache=true

File: crates/rspack_core/src/compiler/queue.rs [hunk 3]
```diff
@@ -222,6 +222,26 @@ impl AddTask {
...
+        self.original_module_identifier,
+        self.dependencies,
+        *issuer,
+      )?;
+
+      return Ok(TaskResult::Add(Box::new(AddTaskResult::ModuleReused {
+        module: self.module,
+      })));
+    }
+
```
Comment (LingyuCoder) [inline crates/rspack_core/src/compiler/queue.rs:244]: proxy self module to reuse parent module, then the exports info should be valid

**Review by IWANABETHATGUY — APPROVED** [review]

**Comment by JSerFeng** [comment]
!bench

**Pull request merged**